packages/api/src/mcp/MCPServersRegistry.ts

```diff
@@ -42,21 +50,48 @@ export class MCPServersRegistry {
 
     const serverNames = Object.keys(this.parsedConfigs);
 
-    await Promise.allSettled(serverNames.map((serverName) => this.gatherServerInfo(serverName)));
+    await Promise.allSettled(
+      serverNames.map((serverName) => this.initializeServerWithTimeout(serverName)),
+    );
+  }
 
-    this.setOAuthServers();
-    this.setServerInstructions();
-    this.setAppServerConfigs();
-    await this.setAppToolFunctions();
+  /** Wraps server initialization with a timeout to prevent hanging */
+  private async initializeServerWithTimeout(serverName: string): Promise<void> {
+    let timeoutId: NodeJS.Timeout | null = null;
 
-    this.connections.disconnectAll();
+    try {
+      await Promise.race([
+        this.initializeServer(serverName),
+        new Promise<never>((_, reject) => {
+          timeoutId = setTimeout(() => {
+            reject(new Error('Server initialization timed out'));
+          }, this.initTimeoutMs);
+        }),
+      ]);
+    } catch (error) {
+      logger.warn(`${this.prefix(serverName)} Server initialization failed:`, error);
+      throw error;
+    } finally {
+      if (timeoutId != null) {
+        clearTimeout(timeoutId);
+      }
+    }
   }
 
-  /** Fetches all metadata for a single server in parallel */
-  private async gatherServerInfo(serverName: string): Promise<void> {
+  /** Initializes a single server with all its metadata and adds it to appropriate collections */
+  private async initializeServer(serverName: string): Promise<void> {
+    logger.info(`${this.prefix(serverName)} Initializing server`);
+    const start = Date.now();
+
+    const config = this.parsedConfigs[serverName];
+    if (config == null) {
+      logger.warn(`${this.prefix(serverName)} Server config not found`);
+      return;
+    }
```

Comment:
Nit: Since `const serverNames = Object.keys(this.parsedConfigs);` this scenario would never happen. We should declare `keyof this.parsedConfigs` typing for severName to avoid having to do this type check. It doesn't hurt to have this check, sure, but we don't want the code to be bloated with not-null assertions while Typescript is perfectly cable of covering this for us.